Enable CouchDB Web UI and upgrade to 2.3.1

stepanstipl opened the pull request:
This PR addresses [GPII-4023: Enable CouchDB Web UI access](https://issues.gpii.net/browse/GPII-4023).

Changes introduced:

- Enables CouchDB Web UI access by enabling auth prompt for web browser access
- Updates CouchDB version to 2.3.1
  - this update includes mostly minor fixes, no required actions mentioned in changelog
  - see https://docs.couchdb.org/en/latest/whatsnew/2.3.html#version-2-3-1 for details
- Adds rake tasks for comfortable CouchDB access
- Creates a temporary account for UI access

This PR depends on gpii-ops/exekube#63.

I have tested updating existing cluster and didn't observe any issues with the new 2.3.1 version.

Output from the `couchdb_ui` rake task will be:

```console
$ rake couchdb_ui
docker-compose run --rm -p 5984:5984 xk rake couchdb_ui
gcloud config set project $TF_VAR_project_id
...
...
couchdb_ui.sh: Starting port-forwarding
couchdb_ui.sh: Waiting for CouchDB port to be available...
couchdb_ui.sh: Waiting for CouchDB port to be available...
couchdb_ui.sh: Waiting for CouchDB port to be available...
couchdb_ui.sh: Creating temporary UI user (ui)
couchdb_ui.sh:
couchdb_ui.sh: CouchDB port is now being forwarded to your local machine, port 35984.
couchdb_ui.sh:
couchdb_ui.sh: You can access CouchDB Web UI at:
couchdb_ui.sh: http://ui:I2YIX9cEeXktYFbeIjL2Qw1aUBGa2DEW@localhost:35984/_utils
couchdb_ui.sh:
couchdb_ui.sh: You can close this terminal and port-forwarding by pressing ENTER (new line).
couchdb_ui.sh:

couchdb_ui.sh: Removing temporary UI user (ui)
couchdb_ui.sh: Stopping port-forwarding
couchdb_ui.sh: Done
```

Printing out the full URL with CouchDB username and password is probably slightly controversial. I've considered alternatives (documenting a way how to pull the secrets from inside the container, most likely saving secrets to file and copying it outside the docker via `docker cp`) and I believe this is actually safer as it doesn't leave any files behind anywhere on a local machine, and definitely more convenient. 
_Update:_ The printed credentials are only temporary, specifically created for UI access

Note: This approach should work even after CouchDB is migrated to Istio.

stepanstipl commented:
> I know that the developers had an instance of CouchDB installed in their locals for the day to day development, perhaps the use of the default port can conflict these running instances. Could you set a different port to make the experience of using the Cluster's Fauxton more comfortable?

Thanks @amatas, yep I definitely can, makes sense. Any suggestions? What about smth like `35984`

amatas commented:
> What about smth like `35984`

That port might work fine :+1:

natarajaya commented on shared/rakefiles/xk_util.rake:
```diff
@@ -263,4 +263,19 @@ task :restore_snapshot_from_image_file, [:snapshot_files] => [@gcp_creds_file, :
+    echo \"CouchDB port is now being forwarded to your local machine, port 35984.\"
+    echo
+    echo \"You can access CouchDB Web UI at:\"
+    echo \"http://${TF_VAR_secret_couchdb_admin_username}:${TF_VAR_secret_couchdb_admin_password}@localhost:35984/_utils\"
```
Don't like the idea of exposing plaintext credentials here. Maybe limit this task to only dev environments?

stepanstipl replied:
We already have the `RAKE_REALLY_RUN_IN_PRD` protection, so I don't want to introduce another one really (also only admins would be able to run this in prod).

But I agree that exposing the credentials is not nice (and alternatives of copying those around do not seem better), and though if improving this story by generating short-lived credentials just for UI access, I'll look into this.

natarajaya commented on gcp/modules/couchdb/values.yaml:
```diff
@@ -40,6 +40,8 @@ couchdbConfig:
     # http://docs.couchdb.org/en/stable/config/auth.html#authentication-configuration
     require_valid_user: true
+  httpd:
+    WWW-Authenticate: "Basic realm=\"CouchDB\""
```
Security requirement from GPII-3341 says: "HTTP endpoint should not be exposed and should be disabled". Set this only for dev environments?

stepanstipl replied:
This does not expose any new port, nor endpoints. It's still very same port (`5984`) and endpoints that are currently used by apps and are accessible by direct http call when you supply the credentials, the only difference is that this setting will trigger auth popup when accessed in the browser.

stepanstipl commented on shared/rakefiles/scripts/couchdb_ui.sh:
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env sh
```
Thanks for catching this 👍 , this file should not have been committed, my mistake.

stepanstipl commented:
Thanks for reviews @amatas, @natarajaya - I've hopefully addressed all the comments. Additional changes:
- Change the exposed port to `35984` to avoid clashes with default port on devs machines
- Moved the code to the external shell script (as it's more complex now, shell-checked)
- Create a temporary account for UI access (to improve exposing credentials story)

natarajaya commented:
> Create a temporary account for UI access (to improve exposing credentials story)

@stepanstipl Cool, this adds complexity, but I definitely like this more than exposing root username / password. What happens if port forwarding container gets terminated before cleanup part can finish?

stepanstipl commented:
> @stepanstipl Cool, this adds complexity, but I definitely like this more than exposing root username / password. What happens if port forwarding container gets terminated before cleanup part can finish?

Thanks @natarajaya. If the container is terminated before the cleanup part, the `ui` account will stay in place. Its password will get overwritten on the next run (in the `curl` `PUT` call) and it will be cleaned up if that run is left to finish.

natarajaya commented:
> Its password will get overwritten on the next run (in the curl PUT call) and it will be cleaned up if that run is left to finish.

Sounds good! LGTM!
